lib/Dialect/FIRRTL/Transforms/LegacyWiring.cpp

```diff
@@ -0,0 +1,120 @@
+//===- LegacyWiring- legacy Wiring annotation resolver ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the legacy Wiring annotation resolver.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace circt;
+using namespace firrtl;
+
+/// Consume SourceAnnotation and SinkAnnotation, storing into state
+LogicalResult circt::firrtl::applyWiring(const AnnoPathValue &target,
+                                         DictionaryAttr anno,
+                                         ApplyState &state) {
+  auto *context = anno.getContext();
+  ImplicitLocOpBuilder builder(target.ref.getOp()->getLoc(), context);
+
+  // Convert target to Value
+  SmallVector<Value> targetsValues;
+  if (auto portTarget = target.ref.dyn_cast<PortAnnoTarget>()) {
+    auto portNum = portTarget.getImpl().getPortNo();
+    if (auto module = dyn_cast<FModuleOp>(portTarget.getOp())) {
+      builder.setInsertionPointToEnd(module.getBodyBlock());
+      targetsValues.push_back(getValueByFieldID(
+          builder, module.getArgument(portNum), target.fieldIdx));
+    } else if (auto ext = dyn_cast<FExtModuleOp>(portTarget.getOp())) {
+      if (target.instances.empty()) {
+        auto paths = state.instancePathCache.getAbsolutePaths(ext);
+        if (paths.size() > 1) {
+          mlir::emitError(state.circuit.getLoc())
+              << "cannot resolve a unique instance path from the "
+                 "external module target"
+              << target.ref;
+          return failure();
+        }
+        auto inst = cast<InstanceOp>(paths[0].back());
+        builder.setInsertionPointAfter(inst);
+        targetsValues.push_back(getValueByFieldID(
+            builder, inst->getResult(portNum), target.fieldIdx));
+      } else {
+        auto inst = cast<InstanceOp>(target.instances.back());
```

Comment:
minor: Consider sinking the last 2-3 lines after the if? (nearly same code in `else` as at end of `if` true block).